R/TaskClassif_melanoma.R

```diff
@@ -0,0 +1,128 @@
+#' @title Melanoma Image classification
+#' @name mlr_tasks_melanoma
+#' @description
+#' Classification of melanoma tumor images.
+#'
+#' The data comes from the 2020 SIIM-ISIC challenge.
+#'
+#' @section Construction:
+#' ```
+#' tsk("melanoma")
+#' ```
+#'
+#' @template task_download
+#'
+#' @source
+#' \url{https://challenge2020.isic-archive.com/}
+#'
+#' @section Properties:
+#' `r rd_info_task_torch("melanoma", missings = FALSE)`
+#'
+#' @references
+#' `r format_bib("melanoma2021")`
+#' @examplesIf torch::torch_is_installed()
+#' task = tsk("melanoma")
+#' task
+NULL
+
+# @param path (`character(1)`)\cr
+#   The cache_dir/datasets/melanoma folder
+constructor_melanoma = function(path) {
+  base_url = "https://huggingface.co/datasets/carsonzhang/ISIC_2020_small/resolve/main/"
+
+  compressed_tarball_file_name = "hf_ISIC_2020_small.tar.gz"
+  compressed_tarball_path = file.path(path, compressed_tarball_file_name)
+  curl::curl_download(paste0(base_url, compressed_tarball_file_name), compressed_tarball_path)
+  utils::untar(compressed_tarball_path, exdir = path)
+  file.remove(compressed_tarball_path)
+
+  training_metadata_file_name = "ISIC_2020_Training_GroundTruth_v2.csv"
+  training_metadata = data.table::fread(here::here(path, training_metadata_file_name))
+
+  test_metadata_file_name = "ISIC_2020_Test_Metadata.csv"
+  test_metadata = data.table::fread(here::here(path, test_metadata_file_name))
+
+  training_metadata = training_metadata[, split := "train"]
+  test_metadata = setnames(test_metadata,
+    old = c("image", "patient", "anatom_site_general"),
+    new = c("image_name", "patient_id", "anatom_site_general_challenge")
+  )[, split := "test"]
+  metadata = rbind(training_metadata, test_metadata, fill = TRUE)
+
+  melanoma_ds_generator = torch::dataset(
+    initialize = function() {
+      self$.metadata = metadata
+      self$.path = path
+    },
+    .getitem = function(idx) {
+      force(idx)
+
+      x = torchvision::base_loader(file.path(self$.path, paste0(self$.metadata[idx, ]$file_name)))
+      x = torchvision::transform_to_tensor(x)
+
+      return(list(x = x))
+    },
+    .length = function() {
+      nrow(self$.metadata)
+    }
+  )
+
+  melanoma_ds = melanoma_ds_generator()
+
+  dd = as_data_descriptor(melanoma_ds, list(x = c(NA, 3, 128, 128)))
+  lt = lazy_tensor(dd)
+
+  return(cbind(metadata, data.table(image = lt)))
```

Comment:
The return value of this is cached. I don't think we should cache the `lazy_tensor` itself, as we might change some things in the implementation in newer versions. I would only use standard R types as a return value here. The lazy tensor can then be created after the caching. It's not so expensive

Reply:
you can check tiny imagenet for how it's done

Reply:
I think this fixed now. `constructor_melanoma` returns a `data.table()` with file names. And the `cached_constructor()` function in `load_task_melanoma()` constructs the lazy tensor